### ISSUE TITLE
Shingles not working in ES 5.2.1 for search queries with odd no. of words

### ISSUE THREAD

**Issue by moyukhbera** [elastic/elasticsearch#23346]
When trying to search with odd no. of words with my shingle analyzer in ES 5.2.1, no result shows up.

eg: '**new delhi abcd**' does not return any results but '**new delhi abcd xyz**' returns

My analyzers look like below
```
{
    "search": {
        "settings": {
            "index": {
                "analysis": {
                    "filter": {
                        "my_shingle_filter": {
                            "max_shingle_size": "2",
                            "min_shingle_size": "2",
                            "output_unigrams": "false",
                            "type": "shingle"
                        },
                        "autocomplete_filter": {
                            "type": "edge_ngram",
                            "min_gram": "1",
                            "max_gram": "40"
                        }
                    },
                    "analyzer": {
                        "my_shingle_analyzer": {
                            "filter": [
                                "lowercase",
                                "my_shingle_filter"
                            ],
                            "type": "custom",
                            "tokenizer": "whitespace"
                        },
                        "shingle_with_autocomplete": {
                            "filter": [
                                "lowercase",
                                "my_shingle_filter",
                                "autocomplete_filter"
                            ],
                            "type": "custom",
                            "tokenizer": "whitespace"
                        }
                    }
                }
            }
        }
    }
}
```

The mapping for the type is shared below

```
{
    "search": {
        "mappings": {
            "address": {
                "properties": {
                    "full_address": {
                        "type": "text",
                        "norms": false,
                        "fields": {
                            "edge_n_grams": {
                                "type": "text",
                                "norms": false,
                                "analyzer": "autocomplete"
                            },
                            "shingles": {
                                "type": "text",
                                "norms": false,
                                "analyzer": "shingle_with_autocomplete"
                            },
                            "synonym": {
                                "type": "text",
                                "norms": false,
                                "analyzer": "synonym_autocomplete"
                            }
                        },
                        "analyzer": "whitespace"
                    }
                }
            }
        }
    }
}
```

**Query 1**

```
GET search/address/_search?_source=full_address
{
    "query": {
        "match": {
            "full_address.shingles": {
                "query": "new delhi",
                "analyzer": "my_shingle_analyzer"
            }
        }
    }
}
```

**Results**

```
{
    ...
    "hits": {
        "total": 21801,
        "max_score": 8.015874,
        "hits": [{
                "_index": "search",
                "_type": "address",
                "_id": "581c50297fd84ecc35420570",
                "_score": 8.015874,
                "_source": {
                    "full_address": "new delhi nagar palika adarsh vidyalay new delhi nagar palika adarsh vidyalay tilak lane tilak marg area new delhi delhi 110001"
                }
            }, {
                "_index": "search",
                "_type": "address",
                "_id": "581c502a7fd84ecc35422010",
                "_score": 7.013889,
                "_source": {
                    "full_address": "kingdom hall of jehovant witness gyan bharti public school saket new delhi new delhi delhi 110017"
                }
            },
            ...
        ]
    }
}
```

**Query 2**

```
GET search/address/_search?_source=full_address
{
    "query": {
        "match": {
            "full_address.shingles": {
                "query": "new delhi nag",
                "analyzer": "my_shingle_analyzer"
            }
        }
    }
}
```

**Results**

```
{
  ...,
  "hits": {
    "total": 0,
    "max_score": null,
    "hits": []
  }
}
```

Query 2 returns results in ES 5.0.0

**Comment by ywelsch** [comment]
Can you at least link to the discuss issue (which just links to stack overflow)?

https://discuss.elastic.co/t/shingles-not-working-with-odd-no-of-search-words/76356

I've tested this with ES 5.1.2 and ES 5.2.1. The former ES version correctly returns the documents, the latter does not.

The _analyze endpoint returns exactly the same output on both 5.1.2 and 5.2.1 but explain returns the following on 5.1.2:

```
{
  "_index": "search",
  "_type": "address",
  "_id": "581c502a7fd84ecc35422010",
  "matched": true,
  "explanation": {
    "value": 0.25069216,
    "description": "sum of:",
    "details": [
      {
        "value": 0.25069216,
        "description": "sum of:",
        "details": [
          {
            "value": 0.25069216,
            "description": "weight(full_address.shingles:new delhi in 0) [PerFieldSimilarity], result of:",
            "details": [
              {
                "value": 0.25069216,
                "description": "score(doc=0,freq=2.0 = termFreq=2.0\n), product of:",
                "details": [
                  {
                    "value": 0.18232156,
                    "description": "idf(docFreq=2, docCount=2)",
                    "details": []
                  },
                  {
                    "value": 1.375,
                    "description": "tfNorm, computed from:",
                    "details": [
                      {
                        "value": 2,
                        "description": "termFreq=2.0",
                        "details": []
                      },
                      {
                        "value": 1.2,
                        "description": "parameter k1",
                        "details": []
                      },
                      {
                        "value": 0,
                        "description": "parameter b (norms omitted for field)",
                        "details": []
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      },
      {
        "value": 0,
        "description": "match on required clause, product of:",
        "details": [
          {
            "value": 0,
            "description": "# clause",
            "details": []
          },
          {
            "value": 1,
            "description": "*:*, product of:",
            "details": [
              {
                "value": 1,
                "description": "boost",
                "details": []
              },
              {
                "value": 1,
                "description": "queryNorm",
                "details": []
              }
            ]
          }
        ]
      }
    ]
  }
}
```

but the following on 5.2.1:

```
{
  "_index": "search",
  "_type": "address",
  "_id": "581c502a7fd84ecc35422010",
  "matched": false,
  "explanation": {
    "value": 0,
    "description": "Failure to meet condition(s) of required/prohibited clause(s)",
    "details": [
      {
        "value": 0,
        "description": "no match on required clause (MatchNoDocsQuery[\"Matching no documents because no terms present.\"])",
        "details": [
          {
            "value": 0,
            "description": "Matching no documents because no terms present.",
            "details": []
          }
        ]
      },
      {
        "value": 0,
        "description": "match on required clause, product of:",
        "details": [
          {
            "value": 0,
            "description": "# clause",
            "details": []
          },
          {
            "value": 1,
            "description": "*:*, product of:",
            "details": [
              {
                "value": 1,
                "description": "boost",
                "details": []
              },
              {
                "value": 1,
                "description": "queryNorm",
                "details": []
              }
            ]
          }
        ]
      }
    ]
  }
}
```

@jpountz can you have a look?

**Comment by jimczi** [comment]
This is a bug indeed. The shingle filter creates disconnected graph that the Lucene query builder is not able to parse. I've created https://issues.apache.org/jira/browse/LUCENE-7708 to track the progress.

**Comment by ywelsch** [comment]
THanks @jimczi !

**Comment by moyukhbera** [comment]
Thanks @ywelsch @jimczi !

**Comment by colings86** [comment]
@jimczi it looks like the above issue was fixed in Lucene 6.5, should this be an adoptme now?

**Comment by jimczi** [comment]
No further work is needed, closing.
This issue is fixed in Lucene 6.5 and ES 5.4 has been upgraded to this version.
